chore: added roles

thepatrickniyo opened the pull request:
## What does this PR do
- It implements the roles feature
- It adds the feature tests

## Description
- Admin is able to perform the CRUD operations on roles
- Roles can be seeded at once for development purposes by running `npm run db:seeds`

## How to manually test this 
- Check in the branch `chore-roles`
- Run the server
- Open the Swagger API documentation to test the roles

## Background context
- The Create, Update, and Delete endpoints are protected so they can only be accessed by the admin

## Relevant Trello story 
[9-create-roles-and-set-permissions](https://trello.com/c/a6wcZ3PE/9-create-roles-and-set-permissions)

## Screenshots
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/54368821/226496960-ccaa45af-0032-4326-b92f-063e652d8d79.png">
<img width="1363" alt="image" src="https://user-images.githubusercontent.com/54368821/226497022-27c50446-6531-470f-8856-c83acb5bea37.png">

thepatrickniyo committed:
chore: added roles

thepatrickniyo committed:
chore: added chore module

thepatrickniyo committed:
chore: rebase develop

thepatrickniyo committed:
bug(db-scripts): fix db scripts failing with es6

thepatrickniyo committed:
rebased develop

thepatrickniyo committed:
chore: added 2 methods

SauveJeanLuc commented on src/api/controllers/admin.controller.js:
```diff
@@ -0,0 +1,34 @@
+import * as _ from "lodash";
```
This may import more modules than you need and increase your bundle size,

you can use this instead: `import { pick } from “lodash”;`

thepatrickniyo replied:
Thanks, @SauveJeanLuc , according to the feedback they gave me in the standup requesting me to remove the sign-in I ended up totally removing all that including this change.

SauveJeanLuc commented on src/api/enums/EHttpStatusCodes.js:
can't we replace this by the standard library `http-status-codes` ?
this might be less error-prone

thepatrickniyo replied:
Good Idea, I addressed it. I have now employed that lib. check the latest commit

thepatrickniyo committed:
chore: added tests

thepatrickniyo committed:
Merge branch 'develop' of https://github.com/atlp-rwanda/rca-phantom-team1-bn into chore-roles

thepatrickniyo committed:
chore: added test secrets

thepatrickniyo committed:
fix secrets

thepatrickniyo committed:
fix: ci-cd

thepatrickniyo committed:
log: env

thepatrickniyo committed:
log: env

thepatrickniyo committed:
added create dotenv

thepatrickniyo committed:
fix tests

thepatrickniyo committed:
chore: added http-status-codes lib

TuyizeeAnastase reviewed (changes requested):
steps for this task for permissions
- first middleware to check user logged in by checking the token and sending user info logged in
- second middleware, receive user and check role(admin, operator...) 
- call the second middleware in routes 
`route for operator restriction('operator') 
route for admin restrict('admin','operator']`
- add permission to all roles endpoints

TuyizeeAnastase commented on src/db/models/role.js:
```diff
@@ -0,0 +1,28 @@
+  }
+  Role.init(
+    {
+      role: DataTypes.STRING,
```
name title

TuyizeeAnastase commented on src/api/controllers/roles.controller.js:
```diff
@@ -0,0 +1,77 @@
+
+export const getRoles = async (req, res, next) => {
+  try {
+    const roles = await models.role.findAll();
```
create new services file in the services folder named it role.services.js and add this code to find roles

TuyizeeAnastase commented on src/api/controllers/roles.controller.js:
```diff
@@ -0,0 +1,77 @@
+export const getRoleById = async (req, res, next) => {
+  const { role } = req.params;
+  try {
+    const roleData = await models.role.findOne({ where: { role } });
```
add in role.services.js

TuyizeeAnastase commented on src/api/controllers/roles.controller.js:
```diff
@@ -0,0 +1,77 @@
+};
+
+export const roleExists = async (role) => {
+  const roleData = await models.role.findOne({ where: { role } });
```
add in role.services.js to findOne role

TuyizeeAnastase commented on src/api/controllers/roles.controller.js:
```diff
@@ -0,0 +1,77 @@
+export const createRole = async (req, res, next) => {
+  const role = req.body;
+  try {
+    const roleData = await models.role.create(role);
```
add create role in role services

TuyizeeAnastase commented on src/api/controllers/roles.controller.js:
```diff
@@ -0,0 +1,77 @@
+  const { id } = req.params;
+  const newRole = req.body;
+  try {
+    const roleData = await models.role.findOne({ where: { id } });
```
add this to role servics

TuyizeeAnastase commented on src/api/controllers/roles.controller.js:
```diff
@@ -0,0 +1,77 @@
+export const deleteRoleById = async (req, res, next) => {
+  const { id } = req.params;
+  try {
+    const roleData = await models.role.findByPk(id);
```
add this to role services

TuyizeeAnastase commented on src/api/controllers/roles.controller.js:
```diff
@@ -0,0 +1,77 @@
+  const { id } = req.params;
+  try {
+    const roleData = await models.role.findByPk(id);
+    if (!roleData) {
+      return res
+        .status(StatusCodes.NOT_FOUND)
+        .json({ message: "Role not found" });
+    }
```
check this using midleware to check if roleExist

TuyizeeAnastase commented on src/api/middlewares/adminCheck.js:
```diff
@@ -0,0 +1,32 @@
+import ERoles from "../enums/ERole";
+import { decodeJwtToken } from "../utils/jwt";
+
+const adminCheck = async (req, res, next) => {
+  try {
+    const authHeader = req.headers.authorization;
+    if (!authHeader) {
+      return res
+        .status(401)
+        .json({ message: "Authorization header not found" });
+    }
+
+    const token = authHeader.replace("Bearer ", "");
+    const decoded = await decodeJwtToken(token);
+    const isAdmin = decoded.roles.includes(ERoles.ADMINISTRATOR);
+    if (!isAdmin) {
+      return res
+        .status(StatusCodes.FORBIDDEN)
+        .json({ message: "You are not authorized to access this resource" });
+    }
+
+    // If the user is an admin, pass control to the next middleware
+    next();
+  } catch (err) {
+    return res
+      .status(StatusCodes.UNAUTHORIZED)
+      .json({ message: "Invalid or expired token", error: err.message });
+  }
+};
```
refactor this into two middleware function
- check user is logged in using authorization and get the user using id from the decoded user and send to the next middleware
- add a second function to restrict by role
    . function receives many roles and checks if logged user role include in 
   refer to this
     ``export const restrictTo = (...roles)=>{
    return (req, res, next) => {
        const role = req.user.role.title
        if (!roles.includes(role)) {
          return res.status(403).json({
            message: "You are not authorized to perform this action",
          });
        }
        next();
      };
}

TuyizeeAnastase commented on src/api/routes/roles.routes.js:
```diff
@@ -0,0 +1,168 @@
+ *       500:
+ *         description: Internal server error
+ */
+roleRouter.get("/", getRoles);
+
+/**
+ * @swagger
+ * /roles:
+ *   post:
+ *     summary: Create a new role
+ *     tags: [Roles]
+ *     security:
+ *       - bearerAuth: []
+ *     requestBody:
+ *       required: true
+ *       content:
+ *         application/json:
+ *           schema:
+ *             type: object
+ *             properties:
+ *               role:
+ *                 type: string
+ *                 description: Name of the role to create
+ *               description:
+ *                 type: string
+ *                 description: Description of the role
+ *               privileges:
+ *                 type: array
+ *                 items:
+ *                   type: string
+ *                   description: Privileges of the role
+ *     responses:
+ *       201:
+ *         description: Returns the created role
```
all action related to roles must done by only administrator
refer to above comments 
call middleware(checkUSerLogged,restrt('admin'),controller)

TuyizeeAnastase commented on src/db/migrations/20230316082830-create-role.js:
```diff
@@ -0,0 +1,41 @@
+        autoIncrement: true,
+        primaryKey: true,
+      },
+      role: {
```
use title not role

thepatrickniyo committed:
Address reviews
Apply protections
Apply permissions
Refactor middlewares
Add role services
Refactor role controller
Add getby id endpoint
Endpoints refactoring

TuyizeeAnastase commented on src/api/middlewares/restrict.middleware.js:
```diff
@@ -0,0 +1,15 @@
+import { StatusCodes } from "http-status-codes";
+
+const restrictTo = (...roles) => {
+  return (req, res, next) => {
+    const role = req.user.role;
+    if (!roles.includes(role.toUpperCase())) {
+      return res
+        .status(StatusCodes.UNAUTHORIZED)
+        .json({ message: "You are not authorized to perform this action" });
+    }
+    next();
+  };
+};
```
add checkUserLoggedIn and restrictTo into same file `protect.middleware.js`

thepatrickniyo replied:
> I can't see checkLoggedIn and restrict middlewares in roles routes

Check in the index, to increase code smell by reducing redundancy those middleware are used on the primary route of roles in index. Check well.

TuyizeeAnastase commented on src/api/routes/roles.routes.js:
```diff
@@ -92,7 +125,7 @@ roleRouter.post("/", adminCheck, validateCreateRole, createRole);
  *         description: Internal server error
  */
-roleRouter.get("/:role", getRoleById);
+roleRouter.get("/id/:id", roleExistsById, getRoleById);
```
why /id/:id

thepatrickniyo replied:
I added two middlewares. one of getting role by Id and another of getting role by title. Technically that's how to get the two separated 

`roles/id/[role-id]`

`roles/title/[role-title]`

TuyizeeAnastase commented on src/api/routes/roles.routes.js:
```diff
@@ -69,21 +75,48 @@ roleRouter.get("/", getRoles);
+ *       500:
+ *         description: Internal server error
+ */
+roleRouter.get("/title/:title", roleExistsByTitle, getRoleByTitle);
```
use the same router to get all roles and check for a req. query.title(return getRoleByTitle) else return all roles

thepatrickniyo replied:
The reason above applys to this one as well as said above

TuyizeeAnastase reviewed (changes requested):
I can't see checkLoggedIn and restrict middlewares in roles routes

thepatrickniyo committed:
address routes

thepatrickniyo committed:
convert title to lower-case

thepatrickniyo committed:
fix tests

thepatrickniyo committed:
chore: rebased develop

TuyizeeAnastase reviewed (approved):
Please ensure that your name and describe the task correctly. This is not a chore, it is a feature
